File: solutions/haskell/secret-handshake/2/src/SecretHandshake.hs
module SecretHandshake (handshake) where

handshake :: Int -> [String]
handshake n
  | n >= 16 = reverse . handshake $ n - 16
  | otherwise = [events !! i | i <- go n 0]
  where
    events = ["wink", "double blink", "close your eyes", "jump"]
    go nn i
      | nn == 0 = []
      | odd nn = i : go (nn `div` 2) (i + 1)
      | otherwise = go (nn `div` 2) (i + 1)

File: solutions/cpp/secret-handshake/2/secret_handshake.cpp
#include "secret_handshake.h"
#include <algorithm>
#include <map>

namespace secret_handshake {

std::vector<std::string> commands(unsigned int handshake) {
  static std::map<unsigned int, std::string> tbl = {
      {{0b0001U, "wink"},
       {0b0010U, "double blink"},
       {0b0100U, "close your eyes"},
       {0b1000U, "jump"}}};
  std::vector<std::string> commands;
  for (const auto &kv : tbl)
    if (handshake & kv.first)
      commands.emplace_back(kv.second);

  if (handshake & 0b10000U)
    std::reverse(begin(commands), end(commands));
  return commands;
}

} // namespace secret_handshake
